src/dotnet/UI.Blazor.App/Components/AudioRecorder/workers/opus-encoder-worker.ts

```diff
@@ -216,44 +222,71 @@ async function stopRecording(): Promise<void> {
-        if (gains.length) {
-            for (let i = 0; i < queue.length; i++) {
-                const samplesBuffer = queue.peekAt(i);
-                const samples = new Float32Array(samplesBuffer);
-                gains[i] = approximateGain(samples);
-            }
-            const speechGain = average(gains.slice(-5));
-            let startIndex = queue.length - 2;
-            while (startIndex > 0) {
-                const gain = gains[startIndex];
-                if (gain < speechGain/20)
-                    break;
-                startIndex--;
-            }
-            let framesToShift = clamp(startIndex - 1, 0, queue.length - 5); // Keep at least 4 frames
-            debugLog?.log(`processQueue(in): gains: `, gains, framesToShift);
-            while (framesToShift-- > 0)            {
-                const samplesBuffer = queue.shift();
-                void encoderWorklet.releaseBuffer(samplesBuffer, rpcNoWait);
+    try {
+        isProcessing = true;
+        if (fade === 'in') {
+            // calculate precise moment when speech starts - VAD may provide signals with delay
+            // we will try to filter out frames with gain significantly lower than most recent
+            const gains = new Float32Array(queue.length).fill(0);
+            if (gains.length) {
+                for (let i = 0; i < queue.length; i++) {
+                    const samplesBuffer = queue.peekAt(i);
+                    const samples = new Float32Array(samplesBuffer);
+                    gains[i] = approximateGain(samples);
+                }
+                const speechGain = average(gains.slice(-5));
+                let startIndex = queue.length - 2;
+                while (startIndex > 0) {
+                    const gain = gains[startIndex];
+                    if (gain < speechGain/20)
+                        break;
+                    startIndex--;
+                }
+                let framesToShift = clamp(startIndex - 1, 0, queue.length - 5); // Keep at least 4 frames
+                debugLog?.log(`processQueue(in): gains: `, gains, framesToShift);
+                while (framesToShift-- > 0)            {
+                    const samplesBuffer = queue.shift();
+                    void encoderWorklet.releaseBuffer(samplesBuffer, rpcNoWait);
+                }
             }
         }
-    }
-    else if (!isVoiceDetected)
-        return;
+        else if (!isVoiceDetected)
+            return;
 
-    // debugLog?.log(`processQueue:`, fade);
-    try {
+        const expectedWindowSizeSamples = 20 * AR.SAMPLES_PER_MS;
+        const expectedWindowSizeBytes = expectedWindowSizeSamples * 4;
+
+        // debugLog?.log(`processQueue:`, fade);
         while (!queue.isEmpty()) {
             const samplesBuffer = queue.shift();
-            const samples = new Float32Array(samplesBuffer);
+            let samples: Float32Array;
+            if (samplesBuffer.byteLength === expectedWindowSizeBytes) {
```

Comment:
This part with optional resampling the same as in [audio-vad-worker.ts](https://github.com/Actual-Chat/actual-chat/pull/2740/files#diff-6abc4adcbc1fa6e154b56d0293a2af82c07faa912cc099406b2e9c8124960a6b)
I would suggest to extract it and reuse.